speechbrain/lobes/models/huggingface_transformers/mbart.py

```diff
@@ -0,0 +1,214 @@
+        tgt = self.custom_padding(
+            tgt, 0, self.model.model.decoder.config.pad_token_id
+        )
+
+        if self.freeze:
+            with torch.no_grad():
+                # dec_out = self.decoder(input_ids=tgt, encoder_hidden_states=src, encoder_attention_mask=encoder_attention_mask).last_hidden_state.detach()
+                if hasattr(self.model.model, "encoder"):
+                    src = self.model.model.encoder(
+                        inputs_embeds=src
+                    ).last_hidden_state.detach()
+                dec_out = self.model.model.decoder(
+                    input_ids=tgt, encoder_hidden_states=src
+                ).last_hidden_state.detach()
+                dec_out = self.model.lm_head(
+                    dec_out
+                ).detach()  # + self.model.final_logits_bias
+                return dec_out
+
+        # dec_out = self.decoder(input_ids=tgt, encoder_hidden_states=src, encoder_attention_mask=encoder_attention_mask).last_hidden_state
+        if hasattr(self.model.model, "encoder"):
+            src = self.model.model.encoder(inputs_embeds=src).last_hidden_state
+        dec_out = self.model.model.decoder(
+            input_ids=tgt, encoder_hidden_states=src
+        ).last_hidden_state
+        dec_out = self.model.lm_head(dec_out)  # + self.model.final_logits_bias
+        return dec_out
+
+    @torch.no_grad()
+    def decode(self, tgt, encoder_out, enc_len=None):
+        """This method implements a decoding step for the transformer model.
+
+        Arguments
+        ---------
+        tgt : torch.Tensor
+            The sequence to the decoder.
+        encoder_out : torch.Tensor
+            Hidden output of the encoder.
+        enc_len : torch.LongTensor
+            The actual length of encoder states.
+        """
+
+        if tgt.dtype not in [torch.long, torch.int64]:
+            tgt = tgt.long()
+
+        tgt_mask = torch.ones(tgt.size(), device=tgt.device)
+
+        output = self.model.model.decoder(
+            input_ids=tgt,
+            encoder_hidden_states=encoder_out,
+            attention_mask=tgt_mask,
+            output_attentions=True,
+        )
+
+        return (
+            self.model.lm_head(output.last_hidden_state),
+            output.cross_attentions[-1],
+        )
+
+    def custom_padding(self, x, org_pad, custom_pad):
+        """This method customizes the padding.
+        Default pad_idx of SpeechBrain is 0.
+        However, it happens that some text-based models like mBART reserves 0 for something else,
+        and are trained with specific pad_idx.
+        This method change org_pad to custom_pad
+
+        Arguments
+        ---------
+        x : torch.Tensor
+          Input tensor with original pad_idx
+        org_pad : int
+          Orginal pad_idx
+        custom_pad : int
+          Custom pad_idx
+        """
+        out = x.clone()
+        out[x == org_pad] = custom_pad
+
+        return out

```

Comment:
I don't understand. Why are you not changing the padding of SpeechBrain instead? It seems to be very computationally expensive.

Reply:
Because the audio part still needs SB standard padding. I didn't find a way to use 2 different SB padding tokens for the source speech and the target text part. I think changing SB padding will apply the same padding for both parts right?

Reply:
yes I see.